### DISCUSSION TITLE
Embedded SQLite migration to ETCD Cluster

### DISCUSSION THREAD

**Discussion by decipher27** [k3s-io/k3s#9432] [Q&A] [answered]
Hi folks, 

We use [vCluster](https://loft.sh/blog/intro-to-vcluster-tutorial/) which is an open source tool to manage Virtual Kubernetes clusters, the vCluster hosts inside a namespace of a Host Cluster [i.e EKS Cluster]. We use k3s as our Distro and it's backed up by embedded SQLite.

Current K3s version: v1.26.12-k3s1
EKS version: 1.26

For using an external datastore need to set the environment variable in K3s configuration for connecting to the DB. Modify the vCluster template as mentioned [here](https://www.vcluster.com/docs/deploying-vclusters/persistence#datastore-options).

We would like to know what will be the best approach to migrate data from Embedded SQLite to and ETCD Cluster.

**Comment by brandond** [comment]
This is covered in the docs:
 https://docs.k3s.io/datastore/ha-embedded#existing-single-node-clusters

**Reply by decipher27** [reply]
So, it's just that easy by restarting the k3s server with the --cluster-init flag? 👀 

Steps:
- Restart the k3s server with `--cluster-init` flag
- Validate data is migrated. 

Questions:
1/ Could you point me to the code where we have defined the migration using --cluster-init flag please? 
2/ Does this work for other external datastores like psql, mysql and managed instances like RDS Postgres, MYSQL?

**Reply by brandond** [reply, marked as answer]
This is for migrating from sqlite to embedded etcd. That is the only supported migration that k3s will do for you. Any other migration between external datastores you will have to manage on your own using 3rd party backup/restore tools.

**Comment by ramank775** [comment]
Hi - i am planning to migrate from single mater to HA setup. I can see it's as simple as adding a `--cluster-init` flag to move to embedded etcd. 

What i am planning 
- Migrated to embedded etcd from sqlite.
- Spinup new etcd cluster
- Take a backup of etcd data
- Restore the backup data to cluster using etcd cli.
- Migrate the master to use etcd cluster.
- If this fail, use the back to keep using the sqlite or embedded etcd only.

From look of it seems not every difficult, but what raise my concern is why no one mentioned it yet?
Is there any challenge someone has experienced with this approach ?

**Reply by brandond** [reply]
Migrating from sqlite to embedded etcd is covered in the comment directly above yours.

https://github.com/k3s-io/k3s/discussions/9432#discussioncomment-8403715

Why are you doing all those extra steps if you just want to go from single node sqlite to ha with embedded etcd?

**Reply by ramank775** [reply]
I want to migrate to external etcd cluster.

**Reply by brandond** [reply]
We don't support using external etcd. You can, and it should work fine, but we don't support it. That's likely why you don't find any docs or community discussion of doing so.

**Comment by ramank775** [comment]
I am assuming you mean migration to external etcd?

**Reply by brandond** [reply]
No, i mean external etcd, period.

It should work, because that's what upstream uses - but we only test with and officially support kine sqlite/postgres/mysql, and embedded etcd.

**Reply by ramank775** [reply]
Sorry to open new thread, i replied via mail and it automatically do so.

Also official documentation do say support for external etcd ```etcd (certified against version 3.5.4)```  here https://docs.k3s.io/datastore?ext-db=etcd

**Reply by brandond** [reply]
I'd have to check with our QA team as to how often that is validated. I know we don't cover it in CI.

**Reply by brandond** [reply]
I confirmed that it is not something that we cover in our regular CI or QA processes. It was most recently validated in June of 2022: https://github.com/k3s-io/k3s/issues/5859#issuecomment-1196070623

Like I said it should work fine, but at the moment I wouldn't recommend it. Why would you want to use external etcd over etcd managed by k3s?

**Reply by ramank775** [reply]
I'm choosing external etcd over k3s-managed etcd primarily due to hardware limitations in my homelab. My setup includes Proxmox server and Couple of Raspberry Pis. External etcd allows me run master nodes on Raspberry Pi without compromising performance. 

My current single-node k3s setup uses SQLite. To migrate to a high-availability cluster, I must transition through embedded etcd, and external etcd provides a path for that migration, as embedded etcd backups can be restored to external etcd.
